File: android/app/src/main/java/market/foodhome/app/navigation/PendingDeepLinkStore.kt
package market.foodhome.app.navigation

class PendingDeepLinkStore(
    private val navigationPolicy: NavigationPolicy,
) {
    private var pendingUrl: String? = null

    @Synchronized
    fun offer(rawUrl: String): Boolean {
        val decision = navigationPolicy.classify(rawUrl)
        if (decision !is NavigationDecision.Internal) return false
        pendingUrl = decision.uri.toASCIIString()
        return true
    }

    @Synchronized
    fun consumeWhenReady(isBridgeReady: Boolean): String? {
        if (!isBridgeReady) return null
        return pendingUrl.also { pendingUrl = null }
    }
}

File: ios/FoodHomeApp/Navigation/PendingDeepLinkStore.swift
import Foundation

final class PendingDeepLinkStore {
    private let navigationPolicy: NavigationPolicy
    private var pendingURL: URL?

    init(navigationPolicy: NavigationPolicy) {
        self.navigationPolicy = navigationPolicy
    }

    @discardableResult
    func offer(_ rawURL: String) -> Bool {
        guard case let .internal(url) = navigationPolicy.classify(rawURL) else { return false }
        pendingURL = url
        return true
    }

    func consumeWhenReady(isBridgeReady: Bool) -> URL? {
        guard isBridgeReady else { return nil }
        defer { pendingURL = nil }
        return pendingURL
    }
}
